### PR TITLE
Implement np.gcd and math.gcd

### PR DESCRIPTION
**Pull request by ivirshup**
This PR adds implementations for `math.gcd` and `np.gcd`. I've implemented a binary GCD algorithm heavily cribbing from [Julia's implementation](https://github.com/JuliaLang/julia/blob/3efdda65d4046a4825353dddc090a6235764fa8f/base/intfuncs.jl#L28-L49). As pointed out [on gitter](https://gitter.im/numba/numba?at=5db1a4d87477946badc0d480), this is different from how [numpy implements gcd](https://github.com/numpy/numpy/blob/master/numpy/core/src/npymath/npy_math_internal.h.src#L688-L697), but is faster:

```python
import numpy as np
from numba import njit

gcd = lambda x, y, out: np.gcd(x, y, out)
cgcd = njit(gcd)

a = np.random.randint(0, 100000, 10000)
b = np.random.randint(0, 100000, 10000)

out_numpy = np.empty_like(a)
out_numba = np.empty_like(a)

# warmup
cgcd(a, b, out_numba)

%timeit gcd(a, b, out_numpy)
# 957 µs ± 3.65 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit cgcd(a, b, out_numba)
# 351 µs ± 1.11 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

np.all(out_numba == out_numpy)
# True
```

The speed advantage decreases for smaller width values, especially unsigned ones. I think a bit of the issue here is that values are being upcast to `uintp`. I'd like this to be more generic, but I'm unsure if there's an elegant way to cast signed values to unsigned values of equivalent width.

I'm a little unsure of what more I have to do to get this working with CUDA and ROC, and would definitely like some advice on that.

I think this is pretty close to ready for review, I'd just like to see that the tests pass before taking off the WIP label.

### CODE REVIEW

**Comment by esc** [comment]
@ivirshup this is starting to look good! Do let us know when you think it is ready and would like a review.

**Comment by ivirshup** [comment]
I think it's about good!

A few questions I had here:

* Do I need to add code for other architectures to be supported?
* My end goal here is adding lcm, but thought that should go in another PR. Should I add that here instead?
* I thought I'd read something in the dev guide about this, but can't find it now. Can I squash the last three commits?

**Comment by stuartarchibald** [comment]
> I think it's about good!

Great, thanks!


> A few questions I had here:
> 
>     * Do I need to add code for other architectures to be supported?

Perhaps not in this PR, but you are welcome to in a follow up PR.

>     * My end goal here is adding lcm, but thought that should go in another PR. Should I add that here instead?

Perhaps put that in another PR? Shorter PRs tend to get through review quicker.

>     * I thought I'd read something in the dev guide about this, but can't find it now. Can I squash the last three commits?

You can do anything you like with history up to the point of first review, from this point onwards please keep history entirely linear (else code review gets really hard as it's not clear what got changed etc).

Thanks!

**Comment by ivirshup** [comment]
I've squashed those commits, this is ready for review.

**Comment by esc** [comment]
@ivirshup thanks, I have queued it for review.

File: numba/typing/npydecl.py [hunk 1]
```diff
@@ -285,6 +285,9 @@ def generic(self, args, kws):
                      "sqrt", "square", "reciprocal",
                      "divide", "mod", "abs", "fabs" ]
 
+if numpy_version >= (1, 15):
```
Comment (stuartarchibald) [inline numba/typing/npydecl.py]: I'm not sure the `math.gcd` impl should be interested in the NumPy version?

Comment (ivirshup) [reply numba/typing/npydecl.py]: Isn't this numpy ufunc registration?

**Comment by ivirshup** [comment]
Heads up, there are three cases per signed int type that arguably should throw overflow errors. However they behave the same as numpy:

```python
from numba import njit, typeof, types
import numpy as np

def typemin(T):
    """Return minimum value for type T."""
    if isinstance(T, types.functions.NumberClass):
        T = T.instance_type
    elif not isinstance(T, types.Type):
        T = typeof(T)
    return T(T.minval)

gcd = njit(lambda x, y: np.gcd(x, y))

minval = typemin(types.int8) # Is -128
zero = types.int8(0)

gcd(minval, minval)
# -128
gcd(minval, zero)
# -128
gcd(zero, minval)
# -128

assert gcd(minval, minval) == np.gcd(minval, minval)
assert gcd(minval, zero) == np.gcd(minval, zero)
assert gcd(zero, minval) == np.gcd(zero, minval)
```

File: numba/typing/npydecl.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/typing/npydecl.py]: heh, yes, I read the `mathdecl` and this change as a single file diff :)

File: numba/targets/mathimpl.py [hunk 2]
```diff
@@ -410,3 +412,65 @@ def degrees_float_impl(context, builder, sig, args):
 def pow_impl(context, builder, sig, args):
     impl = context.get_function(operator.pow, sig)
     return impl(builder, args)
+
+# -----------------------------------------------------------------------------
+
+
+@intrinsic
```
Comment (stuartarchibald) [inline numba/targets/mathimpl.py]: Perhaps put this in `numba.unsafe.numbers` ?

File: numba/targets/mathimpl.py [hunk 3]
```diff
@@ -410,3 +412,65 @@ def degrees_float_impl(context, builder, sig, args):
...
+        return builder.cttz(src, ir.Constant(ir.IntType(1), 0))
+    return src(src), codegen
+
+
+def _unsigned(T):
+    """Convert integer to unsigned integer of equivalent width."""
+    pass
+
+@overload(_unsigned)
+def _unsigned_impl(T):
```
Comment (stuartarchibald) [inline numba/targets/mathimpl.py]: Might this help here?:
```
n [1]: from numba import types                                                                                                 

In [2]: x = types.int32                                                                                                         

In [3]: getattr(types, 'uint{}'.format(x.bitwidth))                                                                             
Out[3]: uint32
```

File: numba/typing/builtins.py [hunk 4]
```diff
@@ -37,10 +37,11 @@ def generic(self, args, kws):
 @infer_global(abs)
 class Abs(ConcreteTemplate):
     int_cases = [signature(ty, ty) for ty in sorted(types.signed_domain)]
+    uint_cases = [signature(ty, ty) for ty in sorted(types.unsigned_domain)]
     real_cases = [signature(ty, ty) for ty in sorted(types.real_domain)]
     complex_cases = [signature(ty.underlying_float, ty)
                      for ty in sorted(types.complex_domain)]
-    cases = int_cases + real_cases + complex_cases
+    cases = int_cases + uint_cases +  real_cases + complex_cases
```
Comment (stuartarchibald) [inline numba/typing/builtins.py]: This change should be tested?

File: numba/targets/mathimpl.py (hunk 2 above)
Comment (ivirshup) [reply numba/targets/mathimpl.py]: `trailing_zeros`, `unsigned`, or both?

File: numba/typing/builtins.py (hunk 4 above)
Comment (ivirshup) [reply numba/typing/builtins.py]: 👍, added here: 0992234975c952ad8033

File: numba/targets/mathimpl.py (hunk 3 above)
Comment (ivirshup) [reply numba/targets/mathimpl.py]: Changed in ed33011573

File: numba/targets/mathimpl.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/targets/mathimpl.py]: `_trailing_zeros`, it's a useful intrinsic

File: numba/targets/mathimpl.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/targets/mathimpl.py]: thanks

File: numba/targets/mathimpl.py (hunk 2 above)
Comment (ivirshup) [reply numba/targets/mathimpl.py]: @stuartarchibald, I've added that (along with `leading_zeros`, since they're useful together).

Comment (ivirshup) [reply numba/targets/mathimpl.py]: @stuartarchibald, I've added that (along with `leading_zeros`, since they're useful together).

**Comment by stuartarchibald** [comment]
Bump close/open to rerun CI.

File: numba/tests/test_unsafe_intrinsics.py [hunk 5]
```diff
@@ -152,3 +153,63 @@ def use_dump_refcount():
...
+        lz = njit(lambda x: leading_zeros(x))
+        tz = njit(lambda x: trailing_zeros(x))
+
+        evens = [2, 42, 126, 128]
+
+        for T in types.unsigned_domain:
+            assert tz(T(0)) == lz(T(0)) == T.bitwidth
+            for i in range(T.bitwidth):
+                val = T(2 ** i)
+                assert lz(val) + tz(val) + 1 == T.bitwidth
```
Comment (stuartarchibald) [inline numba/tests/test_unsafe_intrinsics.py]: Numba uses the `unittest` functions for assertion, i.e. `self.assertEqual(lz(val) + tz(val) + 1, T.bitwidth)`, for consistency please could these tests be adjusted to not use `assert`, thanks.

File: numba/targets/mathimpl.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/targets/mathimpl.py]: Brilliant, thanks for doing that!

File: numba/tests/test_unsafe_intrinsics.py (hunk 5 above)
Comment (ivirshup) [reply numba/tests/test_unsafe_intrinsics.py]: Ah, right. I normally use pytest, and totally forgot about that.

**Comment by ivirshup** [comment]
> Bump close/open to rerun CI.

Should I be worried about that test? My current assumption is that it's unrelated.

**Comment by stuartarchibald** [comment]
> > Bump close/open to rerun CI.
> 
> Should I be worried about that test? My current assumption is that it's unrelated.

Don't worry about it, it's not this patch :)

I also just realised that this new feature support hasn't been added to the docs. Any chance you could please add a line at an appropriate point in here:
https://github.com/numba/numba/blame/ceab10620764d1ff29e0c0e550d1d5975674819b/docs/source/reference/pysupported.rst#L656
and here:
https://github.com/numba/numba/blame/ceab10620764d1ff29e0c0e550d1d5975674819b/docs/source/reference/numpysupported.rst#L310

Many thanks!

**Comment by ivirshup** [comment]
Added! But I put `gcd` under the math ufunc section for the numpy docs.

On a separate note, when is the next release planned for? I'd really like to get a `lcm` implementation using this code in before then.

**Comment by stuartarchibald** [comment]
> Added! But I put `gcd` under the math ufunc section for the numpy docs.

Great, thanks, this is all done, and thanks for your efforts on adding this feature!

> On a separate note, when is the next release planned for? I'd really like to get a `lcm` implementation using this code in before then.

Sounds good, the earlier it appears in the PR queue the more likely it'll make it :) I imagine it'll be similar to this PR so you've a good base to work from?

0.47.0RC1 is scheduled for the 19th of December 2019, we may code freeze a few days ahead of that depending on how stable the builds are looking. For future, the next release info can be found in the [milestones schedule](https://github.com/numba/numba/milestones) it's also discussed/noted at the bottom of core developer meeting notes, which are made public and are here: https://github.com/numba/numba/wiki/Meeting-Minutes, looks like we need to sync the last few weeks @seibert @sklam ?

**Comment by ivirshup** [comment]
Haha, it was a good base! Just started and I think it's good to go once this merges. 

Probably helps that the definition fits on one line 😄.

**Comment by stuartarchibald** [comment]
If you want to submit a PR before this gets merged (we merge in windows due to long standing reasons!), feel free to branch off this branch and submit based on that.

**Pull request merged**